aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+
+    while True:
+
+        n_requests += 1
+        jresponse = await get_batches(
+            billing_project=billing_project, last_batch_id=last_batch_id, token=token
+        )
+
+        if 'last_batch_id' in jresponse and jresponse['last_batch_id'] == last_batch_id:
+            raise ValueError(
+                f'Something weird is happening with last_batch_job: {last_batch_id}'
+            )
+        last_batch_id = jresponse.get('last_batch_id')
+        if not jresponse.get('batches'):
+            logger.error(f'No batches found for range: [{start}, {end}]')
+            return batches
+        for b in jresponse['batches']:
+            # batch not finished or not finished within the (start, end) range
+            if not b['time_completed'] or not b['complete']:
+                skipped += 1
+                continue
+
+            time_completed = parse_hail_time(b['time_completed'])
+            in_date_range = start <= time_completed < end
+
+            if time_completed < start:
+                logger.info(
+                    f'{billing_project} :: Got {len(batches)} batches '
+                    f'in {n_requests} requests, skipping {skipped}'
+                )
+                return batches
+            if in_date_range:
+                batches.append(b)
+            else:
+                skipped += 1
+
+
+async def get_jobs_for_batch(batch_id, token: str) -> list[dict[str, any]]:
+    """
+    For a single batch, fill in the 'jobs' field.
+    """
+    jobs = []
+    last_job_id = None
+    end = False
+    iterations = 0
+
+    async with aiohttp.ClientSession() as session:
+        while not end:
+            iterations += 1
+
+            if iterations > 1 and iterations % 5 == 0:
+                logger.info(f'On {iterations} iteration to load jobs for {batch_id}')
+
+            q = '?limit=9999'
+            if last_job_id:
+                q += f'&last_job_id={last_job_id}'
+            url = HAIL_JOBS_API.format(batch_id=batch_id) + q
+
+            jresponse = await async_retry_transient_get_json_request(
+                url,
+                aiohttp.ClientError,
+                session=session,
+                headers={'Authorization': 'Bearer ' + token},
+            )
+            new_last_job_id = jresponse.get('last_job_id')
+            if new_last_job_id is None:
+                end = True
+            elif last_job_id and new_last_job_id <= last_job_id:
+                raise ValueError('Something fishy with last job id')
+            last_job_id = new_last_job_id
+            jobs.extend(jresponse['jobs'])
+
+    return jobs
+
+
+async def migrate_entries_from_hail_in_chunks(
+    start: datetime,
+    end: datetime,
+    func_get_finalised_entries_for_batch,
+    billing_project: str = None,
+    entry_chunk_size=500,
+    batch_group_chunk_size=30,
+    log_prefix: str = '',
+    dry_run=False,
+) -> int:
+    """
+    Migrate all the seqr entries from hail batch,
```

Comment:
What's meant by migration in this context? Is this actually deleting something from the source?

Reply:
Renamed to "process" instead of migrate - just a bad choice of terminology